### PR TITLE
Gtkorvo updates

### PR DESCRIPTION
**Pull request by kshitij-v-mehta**
Version updates for packages from the GTKorvo suite.

### CODE REVIEW

**Comment by kshitij-v-mehta** [comment]
Thanks,
Not sure how I messed up the cmake args. Fixed that.
Spack picks up the latest version irrespective of the order, but I changed the order anyway for clarity.

**Comment by adamjstewart** [comment]
> Spack picks up the latest version irrespective of the order

Correct, although we prefer things in descending order for clarity.

File: var/spack/repos/builtin/packages/gtkorvo-atl/package.py [hunk 1]
```diff
@@ -35,10 +35,14 @@ class GtkorvoAtl(CMakePackage):
...
     version('2.1', 'b2324ff041bccba127330a0e1b241978')
 
     depends_on('gtkorvo-cercs-env')
 
     def cmake_args(self):
-        args = ["-DENABLE_TESTING=0", "-DENABLE_BUILD_STATIC=STATIC"]
+        if self.spec.satisfies('@2.2:'):
+            args = ["-DENABLE_TESTING=0", "-DBUILD_SHARED_LIBS=OFF"]
+        else:
+            args = ["-DENABLE_TESTING=0", "-DENABLE_BUILD_STATIC=STATIC"]
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/gtkorvo-atl/package.py]: To remove the duplication, can you do:
```python
args = ['-DENABLE_TESTING=0']

if self.spec.satisfies('@2.2:'):
    args.append('-DBUILD_SHARED_LIBS=OFF')
else:
    args.append('-DENABLE_BUILD_STATIC=STATIC')
```
Also, is `-DENABLE_TESTING` needed for running `make test`? If so, you could do:
```python
if self.run_tests:
    args.append('-DENABLE_TESTING=1')
else:
    args.append('-DENABLE_TESTING=0')
```
`self.run_tests` is a Boolean that says whether `spack install` was run with the `--test` flag. If it was, Spack runs `make test` automatically.

File: var/spack/repos/builtin/packages/libffs/package.py [hunk 2]
```diff
@@ -38,16 +38,21 @@ class Libffs(CMakePackage):
...
             branch='master')
+    version('1.5', 'c41c5f5f448b627740deecd695b7bbf8')
     version('1.1.1', 'aa1c8ad5cf35e8cf76735e3a60891509')
     version('1.1', '561c6b3abc53e12b3c01192e8ef2ffbc')
 
-    depends_on('flex')
-    depends_on('bison')
+    depends_on('flex', type='build', when='@:1.1.1')
+    depends_on('bison', type='build', when='@:1.1.1')
+    depends_on('gtkorvo-cercs-env', type='build', when='@:1.1.1')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/libffs/package.py]: Do you know if these are the exact versions where these dependencies were dropped? It would be nice to be specific in case someone adds another versions between 1.1.1 and 1.5.

Comment (eisenhauer) [reply var/spack/repos/builtin/packages/libffs/package.py]: Everything prior to 1.5 requires flex, bison and cercs-env.

File: var/spack/repos/builtin/packages/gtkorvo-atl/package.py (hunk 1 above)
Comment (eisenhauer) [reply var/spack/repos/builtin/packages/gtkorvo-atl/package.py]: Yes, this would likely be a good change...

File: var/spack/repos/builtin/packages/libffs/package.py (hunk 2 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/libffs/package.py]: In that case let's do `when='@:1.4'`

**Comment by eisenhauer** [comment]
A general query:  The flags here hard-code a build of static libraries to accommodate machines like Titan.  Is there a mechanism in Spack for being smarter about this?

**Comment by adamjstewart** [comment]
Of course! It would look something like:
```python
variant('shared', default=True, description='Enable build of shared libraries')

def cmake_args(self):
    args = []

    if '+shared' in self.spec:
        args.append('-DBUILD_SHARED_LIBS=ON')
    else:
        args.append('-DBUILD_SHARED_LIBS=OFF')
```
Unfortunately, I don't know of a way to default to `~shared` on Titan and `+shared` elsewhere. It would have to be done manually from the command-line or in `packages.yaml`. If there is some architecture that specifically describes which machines don't support `+shared`, we could add a `conflicts()` statement.

File: var/spack/repos/builtin/packages/gtkorvo-atl/package.py (hunk 1 above)
Comment (kshitij-v-mehta) [reply var/spack/repos/builtin/packages/gtkorvo-atl/package.py]: `spack install --test=all libevpath` fails during installation of libffs with 
```
92% tests passed, 6 tests failed out of 74

Total Test time (real) =   8.86 sec

The following tests FAILED:
         22 - context_test (Failed)
         23 - context_test2 (SEGFAULT)
         24 - fortran_test (SEGFAULT)
         29 - write_context_output (Failed)
         30 - context_output_unique (Failed)
         52 - CTOUT_context_test_output.bin (Failed)
Errors while running CTest
Makefile:130: recipe for target 'test' failed
make: *** [test] Error 8
```

This is on a local Debian machine.
@eisenhauer Is this known? Is it ok to proceed with this change `if self.run_tests`?

Comment (eisenhauer) [reply var/spack/repos/builtin/packages/gtkorvo-atl/package.py]: No, not expected...  Let me see if I can duplicate on my Ubuntu VM, just because that's the quickest.  (But generally we're doing nightly regression tests on this stuff, so it'd be weird to have a problem on debian or in spack that didn't show up, say, on redhat.  Really this stuff is pretty OS agnostic, with architecture differences being a far bigger problem.)  BTW, I am having to switch machines today to send my MacBook in for service.  I've transfered my VM to a backup, so I'm hoping for little impact, but can't make promises.

Comment (kshitij-v-mehta) [reply var/spack/repos/builtin/packages/gtkorvo-atl/package.py]: Also, how does `-DENABLE_TESTING=0` affect the tests?
I set `-DENABLE_TESTING=0` for all cases and ran `spack install --tests=all libevpath`. 
```
if self.run_tests:
    args.append('-DENABLE_TESTING=0')
else:
    args.append('-DENABLE_TESTING=0')
```

libffs fails at the testing phase even if -DENABLE_TESTING=0.

**Comment by cdwdirect** [comment]
@kshitij-v-mehta 

When I pull this for testing on my Ubuntu (Linux Mint) workstation, without `libffs` or `libevpath` installed, and run `spack install libevpath %gcc@4.9.3` there are with no issues with `libffs`, but `libevpath` still fails:

```
70 errors found in build log:
     136    -- BUILD_TESTING = ON  (options are: ON, OFF)
     137    -- BUILD_SHARED_LIBS = OFF  (options are: ON, OFF, default OFF)
     138    -- EVPATH_TRANSPORT_MODULES = OFF ()
     139    -- Change a value with: cmake -D<Variable>=<Value>
     140    -- -----------------------------------------------------------------------------
     141    -- Configuring done
  >> 142    CMake Error at CMakeLists.txt:508 (add_executable):
     143      Target "cmprobe" links to target "enet::enet" but the target was not found.
     144      Perhaps a find_package() call is missing for an IMPORTED target, or an
     145      ALIAS target is missing?
```

All of the errors are the same: `...links to target "enet::enet" but the target was not found.`

Looking in the root `CMakeLists.txt` for libevpath, it looks like it's being imported globally:

```
253 # enet transport
254 set(RUN_ENET_TESTS FALSE)
255 find_package(enet 1.3.13)
256 if(ENET_FOUND)
257   set(RUN_ENET_TESTS TRUE)
258   if(EVPATH_TRANSPORT_MODULES)
259     add_library(cmenet MODULE cmenet.c ip_config.c)
260     set_target_properties(cmenet PROPERTIES
261       OUTPUT_NAME ${EVPATH_LIBRARY_PREFIX}cmenet)
262 
263     target_link_libraries(cmenet evpath enet::enet)
264     list(APPEND EVPATH_TRANSPORT_TARGETS cmenet)
265   else()
266     target_sources(EVPath PRIVATE cmenet.c)
267     target_link_libraries(EVPath PRIVATE enet::enet)
268     _pkg_get_target_prefix(enet::enet enet_PREFIX)
269     list(APPEND _pkg_config_pfxs "${enet_PREFIX}")
270     list(APPEND _pkg_config_private_reqs "enet >= ${enet_VERSION}")
271   endif()
272 else()
273   message(STATUS " - Enet library was not found.  This is not a fatal error, just that the Enet transport will not be built.")
274 endif()
```

Not sure if it matters, but I noticed that `gtkorvo-enet` hadn't installed automatically, so I ran `spack install gtkorvo-enet %gcc@4.9.3` and `spack load gtkorvo-enet` without issues. Attempting to `spack install libevpath %gcc@4.9.3` after that produced the same 70 error messages shown above.

I have a fully Spack-ified development environment where everything has been built `%gcc@4.9.3` -- it might be time to upgrade to a new gcc, as that may be causing some of these issues?  Let me know if you suspect that is it.

**Comment by adamjstewart** [comment]
> Not sure if it matters, but I noticed that `gtkorvo-enet` hadn't installed automatically, so I ran `spack install gtkorvo-enet %gcc@4.9.3` and `spack load gtkorvo-enet` without issues. Attempting to `spack install libevpath %gcc@4.9.3` after that produced the same 70 error messages shown above.

If `gtkorvo-enet` isn't a dependency, Spack won't build it. Running `spack load` should have no effect, as Spack purifies the build environment by deleting environment variables like `PATH` and `LD_LIBRARY_PATH`. To see a list of dependencies, run `spack spec libevpath`.

**Comment by cdwdirect** [comment]
@kshitij-v-mehta @eisenhauer : I was able to get it to install with the following steps. Perhaps this can help with the debugging:

**When I enabled shared libraries, those 70 identical errors dropped to 2:**

```
2 errors found in build log:
     36    -- BUILD_TESTING = ON  (options are: ON, OFF)
     37    -- BUILD_SHARED_LIBS = ON  (options are: ON, OFF, default OFF)
     38    -- EVPATH_TRANSPORT_MODULES = ON (cmselect;cmsockets;cmudp;cmmulticast;cmib;cmnnti)
     39    -- Change a value with: cmake -D<Variable>=<Value>
     40    -- -----------------------------------------------------------------------------
     41    -- Configuring done
  >> 42    CMake Error at CMakeLists.txt:356 (add_library):
     43      Target "cmnnti" links to target "enet::enet" but the target was not found.
     44      Perhaps a find_package() call is missing for an IMPORTED target, or an
     45      ALIAS target is missing?
     46    
     47    
  >> 48    CMake Error at CMakeLists.txt:356 (add_library):
     49      Target "cmnnti" links to target "enet::enet" but the target was not found.
     50      Perhaps a find_package() call is missing for an IMPORTED target, or an
     51      ALIAS target is missing?
     52    
     53    
     54    -- Generating done
```

I did a `spack stage libevpath` and went to the staged `evpath-4.4.0/CMakeLists.txt` file and commented out the following section:

```
347 CHECK_LIBRARY_EXISTS(ugni GNI_CqCreate "" HAVE_UGNI)
348 
349 # if(HAVE_IBVERBS OR HAVE_UGNI)
350 #   find_package(NNTI)
351 #   if(NNTI_FOUND)
352 #     if(CercsArch STREQUAL "ppc64")
353 #       set(RUN_NNTI_TESTS FALSE)
354 #     endif()
355 #     if(EVPATH_TRANSPORT_MODULES)
356 #       add_library(cmnnti MODULE cmnnti.c)
357 #       set_target_properties(cmnnti PROPERTIES
358 #         OUTPUT_NAME ${EVPATH_LIBRARY_PREFIX}cmnnti)
359 #       target_link_libraries(cmnnti evpath nnti::nnti enet::enet atl::atl)
360 #       list(APPEND EVPATH_TRANSPORT_TARGETS cmnnti)
361 #     else()
362 #       if(NOT BUILD_SHARED_LIBS)
363 #         list(APPEND EXTRA_INSTALL_PACKAGES NNTI)
364 #       endif()
365 #       target_sources(EVPath PRIVATE cmnnti.c)
366 #       target_link_libraries(EVPath PRIVATE nnti::nnti enet::enet atl::atl)
367 #       list(APPEND _pkg_config_private_libs ${NNTI_LIBRARIES})
368 #     endif()
369 #   endif()
370 # endif()
```

Running `spack install --dont-restage libevpath %gcc@4.9.3` and it installed without any errors.

**Comment by eisenhauer** [comment]
If libevpath is setup to somehow prefer an older version of enet, that's a problem.  Basically the newest revs of all these packages represent a departure from prior CMake conventions, target naming, etc.  The newest should be built with the newest.  I guess we should make the dependencies reflect all this for different versions.

**Comment by cdwdirect** [comment]
OK... I lied.  I uninstalled libevpath and made the following changes:

1. Un-commented the section I'd commented out in the libevpath CMakeLists.txt (~line 356)
2. Edited the CMake to look for the newer enet version: `find_package(enet 1.3.14)` (line 255)
3. Edited the libevpath Spack spec to disable shared libraries (the prior default), and added `depends_on('gtkorvo-enet@1.3.14')`

So I put things back how they were when I first checked out this PR, added the versioned dependency on gtkorvo-enet to the spec, and fixed the version number referenced in EVPath's CMake.

**It now installs fine!** :pray:  That version difference is all that was going wrong.

**Comment by eisenhauer** [comment]
Ugh.  OK, I didn't realize specific versions of prior packages were now referenced explicitly in the GTkorvo CMake files.  I did not update those when I rolled the new releases.  Let me try again, which will require new releases, and yet more updates to this for the new versions...  Sorry I didn't get that right the first time.  I won't do this tonight, but before noon tomorrow.

**Comment by eisenhauer** [comment]
OK, maybe understanding a bit more.  Part of the confusion is that enet isn't really a hard dependency of evpath.  If it's found, we build with an enet transport, but if not we don't, so it wasn't included in the dependency list.   Presumably doing this in Spack style, +enet should be a variant, with an explicit dependency on gtkorvo-enet@1.3.14 for evpath 4.4.0.  Something like:
variant('enet-transport', default=True, description='Build an ENET transport for EVpath')
    if '+enet-transport' in self.spec:
       depends_on('gtkorvo-enet@1.3.14')
If this is done, it shouldn't be necessary to change the evpath/CMakeLists.txt.  That version is a lower-bound on acceptable versions, so it should be satisfied with the 1.3.14.  That means that I don't need to reroll new versions of the GTkorvo software stack.  I've tested this hypothesis and it seems to let me install libevpath without error.  What I'm *not* sure of is exactly where I should place the conditional above.  It uses 'self', so it needs to be in a function?  I placed it in cmake_args() for my testing, but I'm not sure that's right and my perusal of the packaging document didn't give me any better ideas.

**Comment by adamjstewart** [comment]
> Presumably doing this in Spack style, +enet should be a variant, with an explicit dependency on gtkorvo-enet@1.3.14 for evpath 4.4.0.

That's correct, although your syntax is a little off. It would look like:
```python
variant('enet_transport', default=True, description='Build an ENET transport for EVpath')

depends_on('gtkorvo-enet@1.3.14', when='+enet_transport')
```
Note that I changed the variant name to use an underscore instead of a dash. Variants can be disabled using `-enet_transport`, so although it wouldn't confuse the parser, `-enet-transport` might confuse a human. Alternatively, `+enet` is also fine.

**Comment by eisenhauer** [comment]
Excellent, @kshitij-v-mehta want to incorporate that into the PR?  BTW, I have been able to reproduce the test failures and will followup with those.  I'm still not sure what's going on there as the relevant code runs without failure elsewhere.  When those are sorted I'll likely need to roll a new FFS release, but we could at least get these changes in and deal with an additional release when it's ready.

**Comment by kshitij-v-mehta** [comment]
evpath depends on enet 1.3.14 for v 4.4.0+. Should it depend on enet 1.3.13 for evpath <4.4.0 ?

**Comment by eisenhauer** [comment]
Yes, that should work.  (There are only two enet releases, so hopefully….)

On Jul 10, 2018, at 11:10 AM, Kshitij Mehta <notifications@github.com<mailto:notifications@github.com>> wrote:


evpath depends on enet 1.3.14 for v 4.4.0+. Should it depend on enet 1.3.13 for evpath <4.4.0 ?

**Comment by eisenhauer** [comment]
Ugh.  The FFS tests seem to run without failure on my Ubuntu VM *outside* of Spack.

**Comment by kshitij-v-mehta** [comment]
`spack install libevpath +enet_transport` worked fine, but
`spack install libevpath@4.2.4 +enet_transport` fails with 

```
==> Error: ProcessError: Command exited with status 2:
    'make' '-j2'

11 errors found in build log:
     231    [ 17%] Linking C executable cmprobe
     232    /usr/bin/cmake -E cmake_link_script CMakeFiles/cmprobe.dir/link.txt --verbose=1
     233    /home/kmehta/codar-spack/lib/spack/env/gcc/gcc   -Wall  -O2 -g -DNDEBUG   CMakeFiles/cmprobe.dir/cmprobe.c.
            o  -o cmprobe  -L/tmp/kmehta/spack-stage/spack-stage-SjfoB_/evpath-4.2.4/spack-build  -L/home/kmehta/codar-
            spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/gtkorvo-atl-2.2-ybdhjbpxgsq2npj44skzgmrajyzmxzwu/lib  -L/hom
            e/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/libffs-1.5-n3qzzd5jr47vo5zjxf22hh3cnovlh7vx/l
            ib  -L/home/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/gtkorvo-dill-2.4-3jhyw5ot5hicrfo6gt
            mls7jxymsmkgeu/lib  -L/home/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/gtkorvo-enet-1.3.13
            -5oyyenchtezgnlnpwqhqm5qh3pz6tvor/lib -Wl,-rpath,/tmp/kmehta/spack-stage/spack-stage-SjfoB_/evpath-4.2.4/sp
            ack-build:/home/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/gtkorvo-atl-2.2-ybdhjbpxgsq2npj
            44skzgmrajyzmxzwu/lib:/home/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/libffs-1.5-n3qzzd5j
            r47vo5zjxf22hh3cnovlh7vx/lib:/home/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/gtkorvo-dill
            -2.4-3jhyw5ot5hicrfo6gtmls7jxymsmkgeu/lib:/home/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0
            /gtkorvo-enet-1.3.13-5oyyenchtezgnlnpwqhqm5qh3pz6tvor/lib -rdynamic libevpath.a /home/kmehta/codar-spack/op
            t/spack/linux-debian9-x86_64/gcc-6.3.0/libffs-1.5-n3qzzd5jr47vo5zjxf22hh3cnovlh7vx/lib/libffs.a /home/kmeht
            a/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/gtkorvo-atl-2.2-ybdhjbpxgsq2npj44skzgmrajyzmxzwu/lib
            /libatl.a /home/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/gtkorvo-dill-2.4-3jhyw5ot5hicrf
            o6gtmls7jxymsmkgeu/lib/libdill.a -lpthread -lm /home/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-
            6.3.0/gtkorvo-enet-1.3.13-5oyyenchtezgnlnpwqhqm5qh3pz6tvor/lib/libenet.a /home/kmehta/codar-spack/opt/spack
            /linux-debian9-x86_64/gcc-6.3.0/libffs-1.5-n3qzzd5jr47vo5zjxf22hh3cnovlh7vx/lib/libffs.a /home/kmehta/codar
            -spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/gtkorvo-atl-2.2-ybdhjbpxgsq2npj44skzgmrajyzmxzwu/lib/libatl
            .a /home/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/gtkorvo-dill-2.4-3jhyw5ot5hicrfo6gtmls
            7jxymsmkgeu/lib/libdill.a -lpthread -lm
     234    [ 18%] Linking CXX executable auto_test
     235    cd /tmp/kmehta/spack-stage/spack-stage-SjfoB_/evpath-4.2.4/spack-build/tests && /usr/bin/cmake -E cmake_lin
            k_script CMakeFiles/auto_test.dir/link.txt --verbose=1
     236    /home/kmehta/codar-spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/libffs-1.5-n3qzzd5jr47vo5zjxf22hh3cnovlh7
            vx/lib/libffs.a(standard.c.o): In function `dlload_externs':
  >> 237    /home/kmehta/codar-spack/var/spack/stage/libffs-1.5-n3qzzd5jr47vo5zjxf22hh3cnovlh7vx/ffs-1.5/cod/standard.c
            :687: undefined reference to `dlopen'
  >> 238    /home/kmehta/codar-spack/var/spack/stage/libffs-1.5-n3qzzd5jr47vo5zjxf22hh3cnovlh7vx/ffs-1.5/cod/standard.c
            :690: undefined reference to `dlsym'
  >> 239    collect2: error: ld returned 1 exit status
```

Looks like test cases are being run even though I did not supply `--tests=all` to `spack install`. Libffs seems to try to open a dynamic library, but all dependencies except enet and cercs-env are built as static libraries.

**Comment by eisenhauer** [comment]
Test cases aren't being run, but they are being built.  The BUILD_TESTING flag is newer and there was no equivalent in evpath 4.2.4; tests were just always built.  

WRT the link failure.  This will be platform dependent.  FFS has the ability to dlopen() data transformation modules on demand.   Modules are similar to dynamic libraries on most platforms, but linking statically doesn't necessarily preclude you from using dlopen(), so the capability isn't automatically turned off just because you asked for static linking.  However, whether or not you need a -ldl flag is platform dependent, and it looks like the old CMake code isn't getting this right for that platform.  Not really sure what to do about that but ignore it.   Fixing things so that they compile and run in more places is part of what the new versions are about...
